### ISSUE TITLE
working 'getting started' guide for windows

### ISSUE THREAD

**Issue by mrpurest** [nerdvegas/rez#469]
My first step when assessing rez for use in our (mostly windows) studio was to go through the getting started guide.

However the steps in the guide do not currently work on windows due to issues already reported.

I understand that there are workarounds, but I don't understand what they are at this stage - I think its important that this is addressed in the documentation so that new users can step through the getting started guide without getting stuck.

**Comment by nerdvegas** [comment]
Could you summarize what those issues are? I don't develop on windows and
have a lot of other issues to track so it isn't easy to be on top of this.
I released a new rez version a month or so back which merges several
Windows-related PRs, and so I also don't know to what extent those do or do
not address the problems you're hitting.

Please also include the rez version you're using in case issues have been
fixed since (though I assume you're working from master or latest?).

Thanks,
A

**Comment by mrpurest** [comment]
I guess this is a duplicate of  #431  in a way.

The rez-build does not work because of cmake issues.   ( see #348 )

The way to get around this seems to be to use the rezbuild.py file rather than CMakeLists.txt
The example hello_world does have a file rezbuild.py.example and this does build without error, however it looks like it has not been tested on windows since you cannot run the 'hello' script with the package afterwards.

We are experimenting at the moment, and the system looks very promising, but we will have to make decisions as to how to use it ourselves before even completing the 'getting started'.

Should we make a 'hello_world_windows' in the example_packages directory?

R

**Comment by nerdvegas** [comment]
Hello,

Please include the rez version you are testing with.

See: https://github.com/nerdvegas/rez/releases/tag/2.17.1

Does this version fix the issue?

Thx
A

**Comment by ISkipy** [comment]
Hello,

Here is the issues that we got:
![capture](https://user-images.githubusercontent.com/35566104/35099215-05222740-fc4f-11e7-99ef-48de4a7bc88e.PNG)

and:
![capture2](https://user-images.githubusercontent.com/35566104/35099222-0dc94c2a-fc4f-11e7-9d8d-79f87fc63d37.PNG)


I understand that as we got an error on the last one we are not able to run this:
![capture3](https://user-images.githubusercontent.com/35566104/35099303-6639f936-fc4f-11e7-979c-b0f828cfc025.PNG)

**Comment by instinct-vfx** [comment]
Disclaimer: I have not looked specifically into this case, so take with a pinch of salt.

I think this is related to the way rez (or more specifically the cmd shell plugin) clears and resets environment variables for resolved contexts. It does not use the environment of the shell it is launched from, but the environment that is persisted in the registry. So even if you run rez-build from a developer shell (or run vcvarsall.bat prior to rez-build) it still won't be able to discover cl.exe and hence cmake will fail.

If you persist the path to cl.exe in your PATH variable rather then setting it in the shell you are getting one step further. It will still fail because there is more magic happening i guess, but this is the root of the issue.

I am using cmake in a quite a few of our rez packages, but i am calling it from rez-build or build_commands because i need some command line features currently not supported.

I am still not sure taking the environment from the registry rather than the current shell is correct or yields expected behavior but so far lacked the time to take a deeper dive.

**Comment by mrpurest** [comment]
So if we agree that the current example package cannot work on windows, I would like to make a package variant / new example package for windows.

I'm talking about a super simple "hello world" package.

I would like some hints as to how I should do this.  do I make a second 'hello_world_windows' package?

R

**Comment by instinct-vfx** [comment]
I think it would actually be better to fix the cmake integration to properly work on windows. That`s quite some task though i am afraid.

**Comment by mrpurest** [comment]
I agree it would be better - but I don't understand enough about rez or cmake or windows to help with that.  
Getting more windows studios using rez would probably help - and a functional getting started guide would help with that.

I don't understand enough about rez to make the windows alternative package either (partly because it was not possible to do the 'getting started') but since this is basically what I will need to learn inorder to use rez I feel like its worth trying.   Perhaps you have an example package you could share?

**Comment by instinct-vfx** [comment]
I don't really have anything ready to share but i can explain my current approach. Note this also contains a custom argument for rez-build to add a --generate switch that allows generating only solution files. Also note, that you need a working cmake setup for this to work of course.

First, in package.py add the following (to prepare for the deprecation of bez and replace it with a custom build command)

`build_command = "python {root}/rezbuild.py {install}"`

For the same reason add the following to the end of rezbuild.py:

```
if __name__ == '__main__':
    build(source_path=os.environ['REZ_BUILD_SOURCE_PATH'],
          build_path=os.environ['REZ_BUILD_PATH'],
          install_path=os.environ['REZ_BUILD_INSTALL_PATH'],
          targets=sys.argv[1:])
```
And in build() i basically do the following:

```
def build(source_path, build_path, install_path, targets):	
    # Generate visual studio project files, note this is for a nuke plugin, so i am using a specific platform target as that is required for the specific nuke version
    subprocess.call(["cmake", "-G", "Visual Studio 14 2015 Win64", "-T", "v100", os.path.join(source_path, "source path"), "-DOUTPUT_PATH=" + build_path], cwd=build_path)

    if int(os.environ["__PARSE_ARG_GENERATE"]):            
        # Return if --generate was given as we only want to generate solution and project files.
        log.info("Solution and project files have been generated in the build folder (" + build_path + ")")
        return

    # Build binaries if --generate was not given
    subprocess.call(["cmake", "--build", ".", "--config", "Release"], cwd=build_path)            

    if "install" not in (targets or []):
        return

    # Copy files to repository (replace with proper setup)
    shutil.copytree(os.path.join(source_path), os.path.join(install_path))
```

**Comment by instinct-vfx** [comment]
P.S. i can try making a corresponding version of hello world with this approach but i am not sure when/if i can find the time.

**Comment by nerdvegas** [comment]
Thorsten it seems like what you have there is *almost* something rez could
support natively.. does the build not work without the preceding call to
`cmake -G ...`? Because what follows just looks like a standard cmake build.

Mrpurest, in terms of what rez can/cannot do build-wise, it may be helpful
to see that rez doesn't take active control of the build process. Mostly
all it does is setup the build environment, set up a bunch of standard
env-vars to tell the build what to do, then runs a build command.

Thorsten's example is slightly more complicated than what you may need, due
to keeping backwards compatibility with a deprecated build tool ('bez').
What you could do instead is simply call a bash script (or .bat, or
whatever) like so:

    # in package.py
    build_command = '{root}/build.bat'

..then in your build.bat you do whatever you like. Build env-vars that get
set are here:
https://github.com/nerdvegas/rez/wiki/Environment-Variables#resolved-build-environment-variables
.

Hth
A

**Comment by mrpurest** [comment]
Thanks for the info.
I am trying to make something simple work at the moment (hello world with some variants).

To be honest I will probably try to bypass the whole 'build' concept.  I don't really understand what the point of it is if I am not compiling anything. 

I still think you should try to make sure it is possible to go through the "getting started" steps - since this is the route people will take to understand what rez does, and in its current (non-working) form it makes rez seem unnecessarily complicated.

**Comment by nerdvegas** [comment]
I don't develop on Windows and so far I've left that support to other
contributors, so I can only say that the Getting Started steps are working
on *nix.

Re building, it doesn't matter if you aren't compiling anything - a build
can be as simple as copying the right files to the right locations, and you
could write a build script to do as such (literally you could just copy
your entire source tree to the build output dir, if that's what you
wanted). Building is an integral part of the workflow and you'll probably
find that it's more hassle to avoid than to use.

Hth
A

**Comment by bpabel** [comment]
I've found that on Windows is far easier to just use the rezbuild.py system and use it to call any other build steps you need to do and then copy the files to the build or install locations.  For python packages, that usually means just calling the `setup.py` script with a custom build location

    python setup.py install --root .\build\py --prefix rez -f

Then you can copy the built python packages from `.\build\py\rez\Lib\site-packages` and `.\build\py\rez\Scripts` to the build paths that you specify in the `package.py` that get added to `PYTHONPATH` and `PATH`.

**Comment by mrpurest** [comment]
Thanks for all your help guys,
I have had a bit of a breakthrough in understanding now and its all starting to make sense.

**Comment by nerdvegas** [comment]
I wonder if supporting setup.py as a native build system might be a good
idea. Clearly there'd have to be a guide about what to include and not
include in the setup.py (for example, it would be nonsensical to include
requirements in the setup.py when in rez they come from package.py).

A

**Comment by bpabel** [comment]
I think better support for setup.py builds would be a good thing, which would certainly also improve rez-pip support.

Most of the the packages I maintain are python libraries and they all essentially use the exact same rezbuild.py file.  [Here's what I'm currently using](https://gist.github.com/bpabel/777a5f04860e3cf28215705a6f0df669).

**Comment by LiamGFX** [comment]
Hi nerdvegas, chrizzFTD, and others,

My name is Liam, I work in a studio, and looking into rez/Windows.

I installed rez 2.20.1 and there are multiple issues with rez/Windows "getting started". The issues are listed below and I'm writing this post with the hope that we can put together a working rez/Windows "getting started".

An open issue is directly below (I'm sure more will pop up), "getting started" steps are after, and my environment is listed at the bottom. Also, I would like to know who, if anyone, uses rez/Windows in "Production" and if anything below can be corrected.

**_Open issue_**:
How are others creating the vs_buildtools package? I haven't found a good solution because I haven't found a way to call vcvars64.bat and have the env the batch file sets up retained.

**_Getting started_**
1) Add 'C:\Python' and 'C:\Python\Scripts' to the PATH env var
2) In 'rez/src/rez/bind/python.py', after Line 69, add:
```
import platform
if 'Windows' == platform.system():
    link += '.exe'
```
3) In 'rez/src/rezplugins/build_system/cmake_files/InstallPython.cmake', change Line 106
from:
`COMMAND ${py_bin} -c 'import py_compile \; py_compile.compile(\"${fabs}\", \"${local_fc}\", None, True)'`
to:
`COMMAND ${py_bin} -c "import py_compile ; py_compile.compile('${fabs}', '${local_fc}', None, True)"`
4) Install rez, add 'C:\rez\Scripts\rez' to the PATH env var, and run rez-bind --quickstart
5) Create a vs_buildtools package and make it dependent on 'os==windows-[VERSION]' and then install the vs_buildtools package
5) Follow https://github.com/nerdvegas/rez/wiki/Getting-Started#building-your-first-package

**_Working environment_**:
- Windows 7 x64
- Python x64 2.7.x
- Microsoft Visual Studio 2017
- PYTHONPATH env var isn't defined
- Initially, PATH env var doesn't include C:\Python or C:\Python\Scripts or C:\rez\Scripts\rez
